Refactor specification to incorporate gittuf design (#16)
adityasaky/gittuf: docs/specification.md

# gittuf Specification

Last Modified: January 19, 2023

Version: 0.1.0

## Introduction

This document describes gittuf, a security layer for Git repositories. gittuf
applies several key properties part of the

> [The Update Framework (TUF)](https://theupdateframework.io/) such as delegations

Reviewer: ```suggestion
[The Update Framework (TUF)](https://theupdateframework.io/) such as delegations,
secure key distribution, key revocation, trust rotation, read / write access control, 
```

and namespaces to Git repositories. This enables owners of the repositories to
distribute (and revoke) contributor signing keys and define policies about which
contributors can make changes to some namespaces within the repository. gittuf

> also implements the Reference State Log which was originally described in an

Reviewer: ```suggestion
also protects against branch state attacks by extending the Reference State Log design
which was originally described in an
```

uses TUF's mechanisms to associate actors with their signing keys. TUF metadata
distributes the public keys of all the actors in the repository and if a key is
compromised, new metadata is issued to revoke its trust.

Second, TUF allows for defining _namespaces_ for the repository. TUF's notion of
namespaces aligns with Git's, and TUF namespaces can be used to reason about
both Git refs and files tracked within the repository. Namespaces are combined
with TUF's _delegations_ to define sets of actors who are authorized to make
changes to some namespace. As such, the owner of the repository can use gittuf
to define actors representing other contributors to the repository, and delegate
to them only the necessary authority to make changes to different namespaces of
the repository.

Policies for gittuf access are defined using a subset of TUF roles. The owners
of the repository hold the keys used to sign the Root role that delegates trust
to the other roles. The top level Targets role and any Targets roles it
delegates to contain restrictions on protected namespaces. The specifics of the
delegation structure vary from repository to repository as each will have its
own constraints.

> A typical TUF delegation connects two TUF Targets role. In the delegations

Reviewer: ```suggestion
A typical TUF delegation connects two TUF Targets roles. In the delegations
```

graph where each node is a Targets role, delegations are the edges connecting
them. The delegations graph is traversed when verifying some target until the

> leaf node is encountered, after which the target's entry is compared against the

Reviewer: This is a little confusing.  There really isn't a meaningful notion of leaf nodes as you may have cycles in delegations and may also have intermediate nodes that directly list targets in addition to delegating.

Author: I'll take another pass. gittuf has a very opinionated approach to TUF delegations. Users are not expected to hand-craft metadata and we always expect the entries to be the commits themselves or the RSL entry so we shouldn't have the intermediate metadata recording hashes issue.

Reviewer: What do you mean that entries are the commits themselves?  I thought they were either a path, branch, or tag?

Author: I've updated this part of the document to try and make it clearer.

Reviewer: I would tweak the new text to more state that TUF's namespacing is general and can be used for various git namespaces like tags, files, branches, etc.

Author: Mentioned namespaces generally and said gittuf applies it to git but modifies the overall workflow to incorporate RSLs etc.
